codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpProtocolUnitTestRequestGenerator.java

```diff
@@ -0,0 +1,238 @@
+                List<HttpRequestTestCase> testCases
+        ) {
+            return new HttpProtocolUnitTestRequestGenerator(model, symbolProvider, protocolName, operation, testCases);
+        }
+    }
+
+    protected HttpProtocolUnitTestRequestGenerator(
+            Model model, SymbolProvider symbolProvider, String protocolName, OperationShape operation,
+            List<HttpRequestTestCase> testCases
+    ) {
+        super(model, symbolProvider, protocolName, operation, testCases);
+    }
+
+
+    @Override
+    protected String unitTestFuncNameFormat() {
+        return "TestClient_$L_$LSerialize";
+    }
+
+    @Override
+    protected Object[] unitTestFuncNameArgs() {
+        return new Object[]{opSymbol.getName(), protocolName};
+    }
+
+    @Override
+    protected void generateTestSetup(GoWriter writer) {
+        writer.openBlock("type queryItem struct {", "}", () -> {
+            writer.write("Key string");
+            writer.write("Value string");
+        });
+    }
+
+    @Override
+    protected void generateTestCaseParams(GoWriter writer) {
+        writer.write("Params $P", inputSymbol);
+        // TODO authScheme
+
+        writer.write("ExpectMethod string");
+        writer.write("ExpectURIPath string");
+
+        writer.write("ExpectQuery []queryItem");
+        writer.write("RequireQuery []string");
+        writer.write("ForbidQuery []string");
+
+        writer.addUseImports(GoDependency.NET_HTTP);
+        writer.write("ExpectHeader http.Header");
+        writer.write("RequireHeader []string");
+        writer.write("ForbidHeader []string");
+
+        writer.write("BodyMediaType string");
+        writer.write("BodyAssert func([]byte) error");
+    }
+
+    @Override
+    protected void generateTestCaseValues(GoWriter writer, HttpRequestTestCase testCase) {
+        writeStructField(writer, "Params", inputShape, testCase.getParams());
+
+        writeStructField(writer, "ExpectMethod", "$S", testCase.getMethod());
+        writeStructField(writer, "ExpectURIPath", "$S", testCase.getUri());
+
+        writeQueryItemsStructField(writer, "ExpectQuery", testCase.getQueryParams());
+        writeStringSliceStructField(writer, "RequireQuery", testCase.getRequireQueryParams());
+        writeStringSliceStructField(writer, "ForbidQuery", testCase.getForbidQueryParams());
+
+        writeHeaderStructField(writer, "ExpectHeader", testCase.getHeaders());
+        writeStringSliceStructField(writer, "RequireHeader", testCase.getRequireHeaders());
+        writeStringSliceStructField(writer, "ForbidHeader", testCase.getForbidHeaders());
+
+        String bodyMediaType = "";
+        if (testCase.getBodyMediaType().isPresent()) {
+            bodyMediaType = testCase.getBodyMediaType().get();
+            writeStructField(writer, "BodyMediaType", "$S", bodyMediaType);
+        }
+        if (testCase.getBody().isPresent()) {
```

Comment:
There's a decent amount of code in here that doesn't vary based on anything being generated for and can live in isolation. I'd recommend moving these out to static files and loading them.

```
IoUtils.readUtf8Resource(getClass(), "protocol-test-empty-comparator.go")
```

Reply:
Ended up moving these to static Go code in the smithy-go module.